foursight_core/react/api/misc_utils.py

```diff
@@ -37,6 +43,72 @@ def get_request_body(request: Request) -> dict:
     return json.loads(request.raw_body.decode())
 
 
+def get_base_url(url: str) -> str:
+    """
+    Returns the base (root) URL if the given URL, i.e. with just the scheme/protocol (e.g.
+    http:// or https://), and the the domain, and port if any. Any credentials (username,
+    password) in the URL (e.g. http:) will NOT be included. For example, given this:
+    https://username:password@example.com/api/react/cgap-supertest/home?arg=12345678
+    this function returns this: https://example.com
+    """
+    try:
+        url = urlparse(url)
+        scheme = url.scheme
+        hostname = url.hostname
+        port = url.port
+        port = '' if (port == 80 and scheme == 'http') or (port == 443 and scheme == 'https') else port
+        return scheme + "://" + hostname + (":" + str(port) if port else "")
+    except Exception:
+        return ""
+
+
+def get_github_url(package: str, file: Optional[str] = None, line: Optional[int] = None) -> Optional[str]:
+    """
+    Returns the GitHub URL to the given package, with the appropriate/current version.
+    Only these packages are supported: foursight-core, foursight-cgap, foursight, dcicutils
+    If the optional file argument is given then returns the versionsed GitHub URL to that file.
+    This is to provide convenience GitHub links to the foursight (e.g. checks) source code in the UI.
+    TODO: Seems like there should be a better more programmatic way to determine this.
+    """
+    github_url = "https://github.com"
+    if package == "foursight-core" or package =="foursight_core":
+        package_source = "foursight-core"
+        package_target = "foursight_core"
+        repo_org = "4dn-dcic"
+    elif package == "foursight-cgap" or package == "chalicelib_cgap":
+        package_source = "foursight-cgap"
+        package_target = "chalicelib_cgap"
+        repo_org = "dbmi-bgm"
+    elif package == "foursight" or package == "chalicelib_fourfront":
+        package_source = "foursight"
+        package_target = "chalicelib_fourfront"
+        repo_org = "4dn-dcic"
+    elif package == "dcicutils" or package == "utils":
+        package_source = "utils"
+        package_target = "dcicutils"
+        repo_org = "4dn-dcic"
+    else:
+        return None
+    try:
+        version = f"v{pkg_resources.get_distribution(package_source).version}"
+    except Exception as e:
+        return None
+    repo_url = f"{github_url}/{repo_org}/{package_source}"
+    if not file:
+        return f"{repo_url}/releases/tag/{version}"
+    if os.path.isabs(file):
+        path = os.path.normpath(file).split(os.sep)
```

Comment:
This is to support a little feature I personally found useful and thought others might as well - there's a link in the UI next to each check which goes directly to the GitHub source for that particular check. This is done by grabbing the (OS) file path for the check function, and this function ends up turning that into a GitHub link. So that's why I thought it appropriate to use os.sep for the file name parsing part (though will only be running on AWS (Linux) servers, except if running locally).